### PR TITLE
174 Implement local watcher

### LINKED ISSUE #174: Implement a local watcher
### Context

This is step 1 of 3 in refactoring the Watcher for IoT use cases (#172). Covers defining the interfaces for watcher and implementing a local watcher.

### Details

The interfaces and functionality of the watching component has been described in proposal [#004](https://github.com/hyperledger-labs/perun-proposals/blob/main/design/004-IoT-Adoption-Watcher.md)

The interface definitions can be placed in a new package `go-perun/watcher` and the local watcher implementation can be placed in `go-perun/watcher/local` package.

### PR DESCRIPTION
**Pull request by manoranjith**
### Overview

- This PR implements the server side component local watcher as per the design described in [proposal#4](https://github.com/hyperledger-labs/perun-proposals/blob/main/design/004-IoT-Adoption-Watcher.md). 
- This component should be initialized by the user of the framework and pass it as an argument to the `client.New` API.

### Implementation details

- The interface definitions are located in a package: `go-perun/watcher`.
- The implementation of local watcher is located in a `go-perun/watcher/local`.

#### Data structures

1. **Watcher**: 
    - This represents the watcher instance, initializes using the `NewWatcher` function.
    - It contains 
        - `channel.RegistererSubscriber` instance for interacting with the blockchain.
        - `registry` that holds a map of all the channels currently registered with the Watcher.
    - Both the above data structures are safe for concurrent use.

2. **ch**:
    - Represents the data corresponding to a channel.
    - It contains
        - `id` of the channel.
        - `params` of the channel. Required while registering dispute for the channel. This will be passed when registering the channel with the `Watcher` via `StartWatching` function.
        - `parent` of this channel. For ledger channels, this is set to `channel.Zero`.

        - `subChsAccess`, (relevant only for ledger channel) a mutex used for concurrency safe access of a family of channels. When a dispute registration is required for a channel, we traverse back to the parent and lock this mutex. It is unlocked only after the dispute registration is completed. This also ensures, we lock only the relevant channels and not the entire registry when registering disputes.

        - `registeredVersion`: when a dispute is registered for a ledger, sub or virtual channel, this field is updated. This is used to ensure, we do not register the same version more than once.
        - `requestLatestTx` and `latestTx` are a set of channels used for retrieving the latest state of this channel without use of mutex locks.

3. **registry**: Used to store and retrieve channels registered with the watcher. It is protected by a mutex.

4. **StatesPubSub** and **AdjudicatorPubSub** implements the functionalities exactly as described in the proposal.
        
#### Methods

1. `NewWatcher` initializes the watcher.
2. `Watcher.StartWatchingForLedgerChannel` and Watcher.StartWatchingForNonLedgerChannel` APIs are used to register a channel with the watcher. Though the implementations of both the APIs are same, two APIs are provided because
    - LedgerChannel does not require parent while,
    - Sub-Channel requires a parent.
    - Regarding the names, I also think we need a better name, used this for time being. 
3. `AdjudicatorSub.Next` can be used by the client for receiving adjudicator events from the watcher.
4. `StatesPub.Publish` can be used by the client to send newer transactions to the watcher.
5. `Watcher.StopWatching` and `Watcher.Shutdown` are yet to implemented.

#### Tests

It was difficult for me to figure out how to add tests using the simulated backend. Particularly, I couldn't figure out how to make the watcher send events to the watcher.

However, I implemented some tests based on generated mocks. This covers both happy path and most of the corner cases. I used these to guide my development. I have added them in the PR (in separate commits). If it is possible, we could replace them with tests using simulated backend or remove them (if they don't fit the testing style used in the project).

#### Resolves #174

### CODE REVIEW

**Review by matthiasgeihs — COMMENTED** [review]
Just a few high level comments. Could you give me and @ggwpez a code intro in our next meeting?

File: watcher/watcher.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	// to the watcher.
+	//
+	// This is initialized when client starts watching for a given channel.
+	// Client can send each state by calling the publish method.
+	//
+	// The publisher will be closed when client requests the watcher to stop
+	// watching or when there is an error. After the client requests the
+	// watcher to stop watching, Publish method must not be called.
+	StatesPub interface {
+		Publish(channel.Transaction) error
```
Comment (ggwpez) [inline watcher/watcher.go:56]: How about a context here?

File: watcher/watcher.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	// from the watcher to the client.
+	//
+	// This is initialized when client starts watching for a given channel.
+	// Client can receive the event via the channel returned by Next method.
+	//
+	// This channel will be closed when client requests the watcher to stop
+	// watching or when there is an error. After the channel is closed, error
+	// message can be read using the Err method.
+	AdjudicatorSub interface {
+		Next() <-chan channel.AdjudicatorEvent
```
Comment (ggwpez) [inline watcher/watcher.go]: Did you have problems with implementing this interface?  
I often think about on how to best do subscriptions in golang.

Comment (ggwpez) [reply watcher/watcher.go]: Please add a comment that this function always returns the same channel.

File: watcher/watcher.go (hunk 1 above)
Comment (ggwpez) [reply watcher/watcher.go:56]: Could you move some of the documentation from the interface to the functions?  
When a user uses the interface he will most likely look at the functions doc.

File: watcher/watcher.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	// latest state on the blockchain.
+	//
+	// Once a channel is closed, client can de-register the channel from the
+	// watcher by calling the StopWatching API. This will close the
+	// AdjudicatorSub and StatesSub. Client must not publish states after
+	// this.
+	Watcher interface {
+		StartWatchingLedgerChannel(context.Context, channel.SignedState) (StatesPub, AdjudicatorSub, error)
+		StartWatchingSubChannel(context.Context, channel.ID, channel.SignedState) (StatesPub, AdjudicatorSub, error)
+		StopWatching(channel.ID) error
```
Comment (ggwpez) [inline watcher/watcher.go:43]: Does this function contain network logic? I think a context would be future proof.

File: watcher/local/statespubsub.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	if s.closed.IsSet() {
+		return s.err
+	}
+	s.err = errors.WithStack(ErrAlreadyClosed)
+	close(s.pipe)
+	s.closed.Set()
+	return nil
+}
+
+func (s *statesPubSub) statesStream() <-chan channel.Transaction {
```
Comment (ggwpez) [inline watcher/local/statespubsub.go:74]: We talked about back-porting the name of this to an earlier commit.

File: watcher/local/registry.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+// retrieveUnsafe retrieves the channel from registry, expecting the caller to
+// have locked the registry.
+func (r *registry) retrieveUnsafe(id channel.ID) (*ch, bool) {
+	ch, ok := r.chs[id]
+	return ch, ok
+}
+
+// retrieve retrieves the channel from registry.
+func (r *registry) retrieve(id channel.ID) (*ch, bool) {
+	r.mtx.Lock()
```
Comment (ggwpez) [inline watcher/local/registry.go:61]: Use the lock function here?

File: watcher/local/registry.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+)
+
+func newRegistry() *registry {
+	return &registry{
+		chs: make(map[channel.ID]*ch),
+	}
+}
+
+// lock locks the registry.
+func (r *registry) lock() {
```
Comment (ggwpez) [inline watcher/local/registry.go:37]: I dont think this function is necessary.

File: watcher/local/watcher.go [hunk 7]
```diff
@@ -15,12 +15,92 @@
...
+	statesPubSub := newStatesPubSub()
+	eventsToClientPubSub := newAdjudicatorEventsPubSub()
+	ch := newCh(id, parent, signedState.Params)
+
+	w.registry.addUnsafe(ch)
+
+	return statesPubSub, eventsToClientPubSub, nil
+}
+
+func newCh(id, parent channel.ID, params *channel.Params) *ch {
```
Comment (ggwpez) [inline watcher/local/watcher.go]: Can you move this to the file that defines `ch`?

File: watcher/local/watcher.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+// limitations under the License.
+
+package local
+
+import (
+	"perun.network/go-perun/channel"
+)
+
+type (
+	ch struct {
```
Comment (ggwpez) [inline watcher/local/watcher.go:58]: You normally name the types self explaining, i think `channel` would make it more understandable.

File: watcher/local/watcher.go [hunk 9]
```diff
@@ -15,12 +15,92 @@
...
+	Config struct {
+		RegisterSubscriber channel.RegisterSubscriber
+	}
+
 	ch struct {
 		id     channel.ID
+		params *channel.Params
 		parent channel.ID
+
+		subChsAccess sync.Mutex
```
Comment (ggwpez) [inline watcher/local/watcher.go:74]: Can you please document when this mutex needs to he locked? So to say, what does it guard?

File: watcher/local/watcher_test.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	ethChannel "perun.network/go-perun/backend/ethereum/channel"
+	_ "perun.network/go-perun/backend/ethereum/channel/test" // For initilizing channeltest
+	"perun.network/go-perun/channel"
+	channeltest "perun.network/go-perun/channel/test"
+	"perun.network/go-perun/pkg/test"
+	"perun.network/go-perun/watcher/internal/mocks"
+	"perun.network/go-perun/watcher/local"
+)
+
+// func init() {
```
Comment (ggwpez) [inline watcher/local/watcher_test.go:40]: Move to `init_test.go` and change log level to `warn` or delete it if you dont need it.

File: watcher/local/watcher.go [hunk 11]
```diff
@@ -104,3 +107,16 @@ func newCh(id, parent channel.ID, params *channel.Params) *ch {
 		parent: parent,
 	}
 }
+
+func (w *Watcher) handleEventsFromChain(eventsFromChainSub channel.AdjudicatorSubscription, eventsToClientPubSub adjudicatorPub,
+	thisCh *ch) {
+	for e := eventsFromChainSub.Next(); e != nil; e = eventsFromChainSub.Next() {
```
Comment (ggwpez) [inline watcher/local/watcher.go:216]: `eventsFromChainSub.Err()` is unused.  
We could also change the `Next()` function to return `(event, error)`  so that correct usage is enforced.

File: watcher/local/watcher.go [hunk 12]
```diff
@@ -25,6 +26,12 @@ import (
 	"perun.network/go-perun/watcher"
 )
 
+const (
+	// Duration for which the watcher will wait for latest transactions
+	// when an adjudicator event is received.
+	statesFromClientWaitTime = 1 * time.Millisecond
```
Comment (ggwpez) [inline watcher/local/watcher.go:33]: I think this should be configurable, what do you think @matthiasgeihs ?  
It would be already good for debugging when using it in deployment.

File: watcher/local/watcher.go [hunk 13]
```diff
@@ -95,28 +106,156 @@ func (w *Watcher) startWatching(ctx context.Context, parent channel.ID, signedSt
...
+			Params: subCh.params,
+			State:  subChTx.State,
+			Sigs:   subChTx.Sigs,
+		}
+	}
+	return subStates
+}
+
+func makeAdjudicatorReq(params *channel.Params, tx channel.Transaction) channel.AdjudicatorReq {
+	return channel.AdjudicatorReq{
```
Comment (ggwpez) [inline watcher/local/watcher.go:299]: Just a comment: I added an entry for adding `NewAdjudicatorReq` in https://github.com/hyperledger-labs/go-perun/issues/179.

File: watcher/local/watcher.go [hunk 14]
```diff
@@ -33,6 +34,16 @@ const (
 )
 
 type (
+	// closer is the interface that wraps the close method.
+	closer interface {
+		close() error
+	}
+
+	// Closer is the interface that wraps the Close method.
+	Closer interface {
```
Comment (ggwpez) [inline watcher/local/watcher.go:43]: Im not really happy with the ad-hoc redefenition of a public `Closer` interface here.  
Please consider using `[pkg/sync] Closer` which also gives so some guarantees as well.

**Review by ggwpez — COMMENTED** [review]
Comments from the meeting.

File: watcher/watcher.go [hunk 15]
```diff
@@ -60,13 +60,14 @@ type (
...
-	// Client can receive the event via the channel returned by Next method.
+	// Client can receive the event via the channel returned by EventStream
+	// method.
 	//
 	// This channel will be closed when client requests the watcher to stop
 	// watching or when there is an error. After the channel is closed, error
 	// message can be read using the Err method.
 	AdjudicatorSub interface {
-		Next() <-chan channel.AdjudicatorEvent
+		EventStream() <-chan channel.AdjudicatorEvent
```
[the remark this answers is not in the document]
Comment (manoranjith) [reply watcher/watcher.go:70]: Updated documentation that `Multiple calls to the EventStream method returns the same channel instance.` in PR #212.

[the remark this answers is not in the document]
Comment (manoranjith) [reply watcher/watcher.go:70]: I think multiplexing is not necessary in our use case, because the only consumer of the events is the watch function.

If at a later point in time, we have the need for consuming the events in multiple places, then we could implement multiplexing.

File: watcher/local/watcher.go [hunk 16]
```diff
@@ -104,3 +107,16 @@ func newCh(id, parent channel.ID, params *channel.Params) *ch {
 		parent: parent,
 	}
 }
+
+func (w *Watcher) handleEventsFromChain(eventsFromChainSub channel.AdjudicatorSubscription, eventsToClientPubSub adjudicatorPub,
+	thisCh *ch) {
+	for e := eventsFromChainSub.Next(); e != nil; e = eventsFromChainSub.Next() {
+		switch e.(type) {
+		case *channel.RegisteredEvent:
```
[the remark this answers is not in the document]
Comment (manoranjith) [reply watcher/local/watcher.go:218]: Updated in #215.

File: watcher/watcher.go (hunk 2 above)
Comment (manoranjith) [reply watcher/watcher.go]: Have added the comment.

Comment (manoranjith) [reply watcher/watcher.go]: Yes, one aspect specific to go-channels is that
1. It is not possible to check if a channel is open, without reading from the channel.
2. Writing to a closed channel will cause panic.

So, the common approach is that, a channel must be closed only by the writer and never by the reader. Also, if there are multiple writers, then they must co-ordinate with each other while closing to ensure no writer writes after the channel has been closed.

Reader can check if a channel is closed by using `data, ok := <-ch` instead of `data := <-ch`. Because, when channel is closed, data will always be zero value and ok will be false.

Comment (manoranjith) [reply watcher/watcher.go]: [Here](https://go101.org/article/channel-closing.html) is a bit long, but very comprehensive article relating to closing of channels. @ggwpez

File: watcher/watcher.go (hunk 3 above)
Comment (manoranjith) [reply watcher/watcher.go:43]: This function will contain network logic (subscribing to blockchain events).

File: watcher/local/statespubsub.go (hunk 4 above)
Comment (manoranjith) [reply watcher/local/statespubsub.go:74]: Yes, I have done that. Now the term `EventStream` is used in `watcher/watcher.go` in the initial commit for that file. See #212.

File: watcher/local/registry.go (hunk 5 above)
Comment (manoranjith) [reply watcher/local/registry.go:61]: I did not get your point ? @ggwpez

File: watcher/local/watcher.go (hunk 12 above)
Comment (manoranjith) [reply watcher/local/watcher.go:33]: Currently, this delay is used only to account for receiving events that are already published, but not read by the watcher. That is also the reason, why this timeout is very small.

However, as we had discussed in the call, in future we could introduce the concept of `Sync interval` and once we do that, we will have to make this as a configurable parameter of the watcher.

File: watcher/local/watcher.go (hunk 13 above)
Comment (manoranjith) [reply watcher/local/watcher.go:299]: I think, in that case, `NewAdjudicatorReq` should be implemented as a function that takes `params` and `Transaction` as parameter and constructs an `AdjudicatorReq` rather than being a method over channel type.

Because,
1. In case of client, it might never need this method.
2. In case of watcher, the watcher will not have access to channel instance, rather only to `params` and `Transaction`.

File: channel/adjudicator.go [hunk 17]
```diff
@@ -25,41 +25,65 @@ import (
...
 	// AdjudicatorEvents. Those events might be triggered by a Register or
 	// Progress call on the adjudicator from any channel participant.
 	Adjudicator interface {
-		// Register should register the given ledger channel state on-chain.
-		// If the channel has locked funds into sub-channels, the corresponding
-		// signed sub-channel states must be provided.
+		Registerer
+		Withdrawer
+		Progresser
+		EventSubscriber
```
[the remark this answers is not in the document]
Comment (manoranjith) [reply channel/adjudicator.go:45]: `EventSubscriber` removed from `Adjudicator`, as it not used by the client anymore. (It is used only by the `Watcher` and `Watcher` uses the `RegisterSubscriber` method.

File: watcher/local/watcher_test.go (hunk 10 above)
Comment (manoranjith) [reply watcher/local/watcher_test.go:40]: Moved to `init_test.go`. See PR #213.

File: watcher/local/watcher.go [hunk 18]
```diff
@@ -15,12 +15,92 @@
...
+	if !ok {
+		return nil, nil, errors.New("parent channel not registered with the watcher")
+	}
+	parentCh.subChsAccess.Lock()
+	defer parentCh.subChsAccess.Unlock()
+	return w.startWatching(ctx, parent, signedState)
+}
+
+func (w *Watcher) startWatching(ctx context.Context, parent channel.ID, signedState channel.SignedState) (
+	watcher.StatesPub, watcher.AdjudicatorSub, error) {
```
[the remark this answers is not in the document]
Comment (manoranjith) [reply watcher/local/watcher.go:114]: Have updated in PR #213.

File: watcher/local/watcher.go (hunk 11 above)
Comment (manoranjith) [reply watcher/local/watcher.go:216]: It was not need here, this PR is not yet handling the closing of adjudicator subscriptions.

I have added that part in a later commit (concerning the StopWatching functionality. See PR #216.

File: watcher/local/watcher.go (hunk 14 above)
Comment (manoranjith) [reply watcher/local/watcher.go:43]: Just noticed the interface itself need not be public. It is just used to store the pub-sub interfaces in the channel instance (of watcher) to access them while closing.

I have renamed this interface. See PR #216.

File: watcher/watcher.go (hunk 1 above)
Comment (manoranjith) [reply watcher/watcher.go:56]: Documenting the method in interfaces was one thing myself and @matthiasgeihs were also discussing  some time ago.

The go's document parser works such that, comment on the interface methods are treated as documentation, it is rather treated as in-line comments for the code.

So, a common practice I noticed in standard library (see [here](https://pkg.go.dev/io#Writer)) was to write documentation on the interfaces rather than methods.

So, we thought we will apply that here.

But, if you see doing otherwise is more beneficial, then we could discuss it and could also change the way we document interfaces.

File: watcher/local/watcher.go [hunk 19]
```diff
@@ -15,12 +15,92 @@
...
+		rs:       cfg.RegisterSubscriber,
+		registry: newRegistry(),
+	}
+	return w, nil
+}
+
+// StartWatchingLedgerChannel starts watching for a ledger channel.
+func (w *Watcher) StartWatchingLedgerChannel(ctx context.Context, signedState channel.SignedState) (
+	watcher.StatesPub, watcher.AdjudicatorSub, error) {
+	return w.startWatching(ctx, channel.Zero, signedState)
```
[the remark this answers is not in the document]
Comment (manoranjith) [reply watcher/local/watcher.go:92]: Have updated the implementation to use `nil` instead of `Channel.Zero` See PRs #213 and #214.

File: watcher/local/registry.go (hunk 5 above)
Comment (ggwpez) [reply watcher/local/registry.go:61]: I will elaborate in #213.

File: watcher/watcher.go (hunk 2 above)
Comment (ggwpez) [reply watcher/watcher.go]: Hm yes the limitations of channels do not make it easier. Somehow every go framework starts re-inventing subscriptions because there is no idiom for that in go.

File: watcher/watcher.go (hunk 3 above)
Comment (ggwpez) [reply watcher/watcher.go:43]: Referenced in #213

File: watcher/local/watcher.go (hunk 8 above)
Comment (ggwpez) [reply watcher/local/watcher.go:58]: moved to #213

File: watcher/local/watcher.go (hunk 9 above)
Comment (ggwpez) [reply watcher/local/watcher.go:74]: moved to #213

File: watcher/local/watcher.go (hunk 13 above)
Comment (ggwpez) [reply watcher/local/watcher.go:299]: Yes i agree, just wanted to let you know that we maybe introduce that function in the future.

File: watcher/local/registry.go (hunk 6 above)
Comment (ggwpez) [reply watcher/local/registry.go:37]: moved to #213

File: watcher/local/watcher.go (hunk 7 above)
Comment (ggwpez) [reply watcher/local/watcher.go]: moved to #213
